### PR TITLE
GTest Integration

### PR DESCRIPTION
**Pull request by ksubramz**
Preliminary GTest support. For more information, please refer to the design document. Thanks Cindy for doing almost all of the work.

### CODE REVIEW

File: framework/test/gtest/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  add_executable(UnitTests main.cpp)
+  target_link_libraries(UnitTests ${GTEST_BOTH_LIBRARIES})
+  target_link_libraries(UnitTests ${Framework_TARGET})
+
+  # This copies the Google Test libraries into the same folder as the
+  # executable so that they can be seen at runtime on Windows.
+  # Mac and Linux use RPATHs and do not need to do this.
+  if (WIN32)
+    foreach( input_file ${GTEST_BOTH_LIBRARIES})
+      string(REPLACE ".lib" ".dll" input_dll_file ${input_file})
```
Comment (jeffdiclemente) [inline framework/test/gtest/CMakeLists.txt]: I'm not sure I understand the purpose of replacing all occurrences of ".lib" with ".dll". If the .lib files aren't meant to be copied over, then just don't copy them.

File: CMakeLists.txt [hunk 2]
```diff
@@ -262,8 +263,20 @@ if(US_BUILD_TESTING)
   enable_testing()
   find_program(US_MEMCHECK_COMMAND valgrind)
   mark_as_advanced(US_MEMCHECK_COMMAND)
+
+  if(GTEST_ROOT STREQUAL "")
+    MESSAGE(STATUS "GTEST_ROOT variable is not defined.")
```
Comment (jeffdiclemente) [inline CMakeLists.txt]: For consistency, can you make the ```MESSAGE``` commands lowercase?

File: framework/test/gtest/CMakeLists.txt (hunk 1 above)
Comment (ksubramz) [inline framework/test/gtest/CMakeLists.txt]: The .lib files aren't getting copied. `GTEST_BOTH_LIBRARIES` is equal to `<path_to_lib>/gtest_main.lib` and `<path_to_lib>/gtest.lib`. Inside the loop, the .lib extension is replaced with .dll. This ensures both the .dll's are copied to the UnitTests executable directory.

Comment (jeffdiclemente) [inline framework/test/gtest/CMakeLists.txt]: Ok. I didn't realize that ```GTEST_BOTH_LIBRARIES``` only had the ```.lib``` files. Is there a CMake variable that has the ```.dll``` files? Taking existing filenames and renaming their file extensions doesn't seem very robust to me.

Comment (ksubramz) [inline framework/test/gtest/CMakeLists.txt]: Unfortunately, no. For the variables exposed by FindGTest package: https://cmake.org/cmake/help/v3.0/module/FindGTest.html

Comment (ksubramz) [inline framework/test/gtest/CMakeLists.txt]: To increase the robustness, I'll see if there are routines to replace only the extension part of a given path.

File: framework/test/gtest/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+  # This copies the Google Test libraries into the same folder as the
+  # executable so that they can be seen at runtime on Windows.
+  # Mac and Linux use RPATHs and do not need to do this.
+  if (WIN32)
+    foreach(lib_fullpath ${GTEST_BOTH_LIBRARIES})
+      get_filename_component(dir ${lib_fullpath} DIRECTORY)
+      get_filename_component(name_no_ext ${lib_fullpath} NAME_WE)
+      set(dll_file "${dir}/${name_no_ext}${CMAKE_SHARED_LIBRARY_SUFFIX}")
+      add_custom_command(TARGET UnitTests POST_BUILD
```
Comment (karthikreddy09) [inline framework/test/gtest/CMakeLists.txt]: GTEST_BOTH_LIBRARIES variable is only set to the DLL paths so why this extra path manipulation ?

File: CMakeLists.txt [hunk 4]
```diff
@@ -262,8 +263,20 @@ if(US_BUILD_TESTING)
   enable_testing()
   find_program(US_MEMCHECK_COMMAND valgrind)
   mark_as_advanced(US_MEMCHECK_COMMAND)
+
+  if(GTEST_ROOT STREQUAL "")
+    message(STATUS "GTEST_ROOT variable is not defined.")
+  else()
+    message(STATUS "GTEST_ROOT: " ${GTEST_ROOT})
+    find_package(GTest REQUIRED)
+endif()
```
Comment (karthikreddy09) [inline CMakeLists.txt]: spacing

File: CMakeLists.txt [hunk 5]
```diff
@@ -262,8 +263,20 @@ if(US_BUILD_TESTING)
...
+
+message(STATUS "Include directories: " ${GTEST_INCLUDE_DIRS})
```
Comment (karthikreddy09) [inline CMakeLists.txt]: "GTest Include directories" is more appropriate for this log message.

File: doc/src/build_instructions.rst [hunk 6]
```diff
@@ -42,6 +42,20 @@ configuration options at hand.
    docs.cppmicroservices.org.
  - **US_BUILD_DOC_MAN** Build the man pages. This is typically only
    enabled on a Unix-like system.
+ - **GTEST_ROOT** Specify the root directory of the Google Test framework
+   installation. These must be specified to build and run unit tests
+   implemented using Google Test. Note that the Google Testing framework must
+   have been built as a shared library.
```
Comment (saschazelzer) [inline doc/src/build_instructions.rst]: Why is it required to be built as a shared library? Shouldn't we be able to support both shared and static, especially when re-using an existing GTest installation?

File: framework/test/gtest/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  include_directories(${GTEST_INCLUDE_DIRS})
+  add_executable(UnitTests main.cpp)
+  target_link_libraries(UnitTests ${GTEST_BOTH_LIBRARIES})
+  target_link_libraries(UnitTests ${Framework_TARGET})
+
+  # This copies the Google Test libraries into the same folder as the
+  # executable so that they can be seen at runtime on Windows.
+  # Mac and Linux use RPATHs and do not need to do this.
+  if (WIN32)
+    foreach(lib_fullpath ${GTEST_BOTH_LIBRARIES})
```
Comment (saschazelzer) [inline framework/test/gtest/CMakeLists.txt]: What if GTest was built as a static library? Then the copying step shouldn't be necessary.

**Comment by jeffdiclemente** [comment]
Hi Sascha,
I'm going to work on updating the gtest integration documentation and responding to your feedback next week.

**Comment by jeffdiclemente** [comment]
I'm just getting back to working on this now. Stay tuned for updates...

**Comment by jeffdiclemente** [comment]
FYI, I'm going to close and abandon this PR in favor of a new one based off of the current development branch since I'm working on the code submission.

I will carry over the existing feedback here and address it in the new PR.

**Comment by jeffdiclemente** [comment]
This PR is replaced by PR #225

**Pull request closed without merging**